eclipse/src/saros/ui/widgets/viewer/project/BaseResourceSelectionComposite.java

```diff
@@ -57,77 +55,300 @@
+  }
+
+  /**
+   * Returns whether the change made by the user was invalid.
+   *
+   * <p>Invalid changes are:
    *
    * <ul>
-   *   <li>If folder or project root is selected, check subtree
-   *   <li>If one (or more) element(s) of project selected, select <b>.classpath</b> and
-   *       <b>.project</b> file as well (if exists)
-   *   <li>If selected resource is not project root, handle gray status of checkboxes in the higher
-   *       tree levels
+   *   <li>A resource being deselected while its parent resource is still selected.
+   *   <li>A parent resource of an already shared resource is selected.
    * </ul>
    *
-   * @param resource
-   * @param checked
+   * @param resource the resource element whose state changed
+   * @param selected the new state of the resource element
+   * @return whether the change is invalid
+   */
+  private boolean isInvalidStateChange(IResource resource, boolean selected) {
+    ISarosSession sarosSession = sessionManager.getSession();
+
+    if (sarosSession != null && resource instanceof IContainer) {
+      Set<IContainer> referencePointParents =
+          getReferencePointParents(sarosSession.getReferencePoints());
+
+      if (referencePointParents.contains(resource)) {
+        return true;
+      }
+    }
+
+    IResource parentResource = resource.getParent();
+
+    boolean parentSelected = checkboxTreeViewer.getChecked(parentResource);
+
+    return parentSelected && !selected;
+  }
+
+  /**
+   * Updates the list of base resources selected by the user.
+   *
+   * <p>If <code>selected=false</code>, the given resource is removed from the list of selected base
+   * resources.
+   *
+   * <p>If <code>selected=true</code>, the given resource is added to the list of selected base
+   * resources. Furthermore, all child resources of the given resources are removed from the list as
+   * they are now represented by the new base resource.
+   *
+   * @param resource the base resource entry changed by the user
+   * @param selected the new state of the resource
    */
-  protected void handleCheckStateChanged(IResource resource, boolean checked) {
-    if (resource instanceof IFile) {
-      this.checkboxTreeViewer.setChecked(resource, checked);
-      setParentsCheckedORGrayed(resource);
-    } else if (resource instanceof IFolder) {
-      this.checkboxTreeViewer.setChecked(resource, checked);
-      this.checkboxTreeViewer.setSubtreeChecked(resource, checked);
-      setParentsCheckedORGrayed(resource);
-    } else if (resource instanceof IProject) {
-      this.checkboxTreeViewer.setGrayChecked(resource, false);
-      this.checkboxTreeViewer.setSubtreeChecked(resource, checked);
+  private void updateSelectedBaseResources(IContainer resource, boolean selected) {
+    if (!selected) {
+      selectedBaseResources.remove(resource);
+
+      return;
     }
+
+    selectedBaseResources.removeIf(
+        selectedBaseResources -> isChildResource(resource, selectedBaseResources));
```

Comment:
Looks pretty weird to use selectedBaseResources again as name. Guess you meant selectedBaseResource.

Reply:
Done.